File: gateway/rabbit/broker.py
import pika
import logging
from functools import partial
from ..config.config import Config


class Broker:
    def __init__(self, config):
        logging.getLogger("pika").setLevel(logging.WARNING)
        self._outputexchange_name = config.outputExchangeName
        self._input_queue_name = config.inputQueueNames[0]

        try:
            self._recvconn = pika.BlockingConnection(
                pika.ConnectionParameters(host="rabbitmq")
            )
            self._sendconn = pika.BlockingConnection(
                pika.ConnectionParameters(host="rabbitmq")
            )

            self._recvchan = self._recvconn.channel()
            self._sendchan = self._sendconn.channel()

            self.inputQueues = self._declare_side(
                self._recvchan,
                config.inputExchangeName,
                config.inputExchangeType,
                config.inputQueueNames,
                config.inputQueueKeys,
            )

            self.outputQueues = self._declare_side(
                self._sendchan,
                self._outputexchange_name,
                config.outputExchangeType,
                config.outputQueueNames,
                config.outputQueueKeys,
            )
        except Exception as e:
            logging.critical(f"Failed to connect with RabbitMQ: {e}")
            raise

    def _init_input(self, chan, config: Config):
        exchange_name = config.inputExchangeName
        q_name = config.inputQueueName
        
        chan.exchange_declare(
            exchange=exchange_name,
            exchange_type="direct",
            durable=True,
        )

        chan.queue_declare(queue=q_name)
        return chan.queue_bind(
            exchange=exchange_name,
            exchange_type="direct",
            durable=True
        )

    def _init_output(self, chan, config: Config):
        exchange_name = config.outputExchangeName
        q_names = config.outputQueueNames
        q_copies = config.outputCopies

        chan.exchange_declare(
            exchange=exchange_name,
            exchange_type="direct",
            durable=True,
        )

        for i in range(len(q_names)):
            q_fmt = q_names[i] + "-{}"
            for id in range(q_copies[i]):
                q_name = q_fmt.format(id)

                chan.queue_declare(queue=q_name)
                q = chan.queue_bind(
                    exchange=exchange_name,
                    exchange_type="direct",
                    durable=True,
                )



    def _declare_side(self, chan, exchange_name, exchange_type, queues_name, queues_keys):
        chan.exchange_declare(
            exchange=exchange_name,
            exchange_type=exchange_type,
            durable=True,
        )

        qs = []
        for i, queue in enumerate(queues_name):
            chan.queue_declare(queue=queue)
            q = chan.queue_bind(
                exchange=exchange_name,
                queue=queue,
                routing_key=queues_keys[i],
            )
            qs.append(q)

        return qs

    def publish(self, routing_key: str, body: str | bytes):
        try:
            self._sendchan.basic_publish(
                exchange=self._outputexchange_name,
                routing_key=routing_key,
                body=body,
            )
        except Exception as e:
            logging.critical(f"Failed to publish message {e}")
            raise

    def consume(self, consumer, callback, conn):
        try:
            self._recvchan.basic_consume(
                queue=self._input_queue_name,
                on_message_callback=partial(callback, conn),
                auto_ack=False,
                exclusive=False,
                consumer_tag=consumer,
            )
            self._recvchan.start_consuming()
        except Exception as e:
            logging.critical(f"Failed to consume message {e}")
            raise
        finally:
            self._recvchan.stop_consuming()

    def close(self):
        if self._recvchan and self._recvchan.is_open:
            try:
                self._recvchan.close()
            except:
                pass
        if self._sendchan and self._sendchan.is_open:
            try:
                self._sendchan.close()
            except:
                pass
        if self._recvconn and self._recvconn.is_open:
            self._recvconn.close()
        if self._sendconn and self._sendconn.is_open:
            self._sendconn.close()

File: workers/rabbit/broker.go
package rabbit

import (
	"fmt"
	"workers/config"

	"github.com/op/go-logging"
	amqp "github.com/rabbitmq/amqp091-go"
)

type Broker struct {
	conn *amqp.Connection
	ch   *amqp.Channel
	con  *config.Config
	log  *logging.Logger
}

func NewBroker(con *config.Config, log *logging.Logger) (*Broker, error) {
	conn, err := amqp.Dial(con.Url)
	if err != nil {
		return nil, err
	}

	ch, err := conn.Channel()
	if err != nil {
		return nil, err
	}

	return &Broker{conn, ch, con, log}, nil
}

func (b *Broker) Init() ([]amqp.Queue, [][]amqp.Queue, error) {
	inputQs, err := b.initInput()
	if err != nil {
		return nil, nil, err
	}

	outputQs, err := b.initOutput()
	if err != nil {
		return inputQs, nil, err
	}

	return inputQs, outputQs, nil
}

func (b *Broker) DeInit() {
	if !b.conn.IsClosed() {
		b.conn.Close()
	}
	if !b.ch.IsClosed() {
		b.ch.Close()
	}
}

func (b *Broker) initInput() ([]amqp.Queue, error) {
	exchangeNames := b.con.InputExchangeNames
	qNames := b.con.InputQueueNames

	qs := make([]amqp.Queue, 0, len(b.con.InputQueueNames))
	for i := range b.con.InputExchangeNames {
		if err := b.exchangeDeclare(exchangeNames[i], "direct"); err != nil {
			return nil, err
		}

		// Build queue name
		nameFmt := qNames[i] + "-%d"
		qName := fmt.Sprintf(nameFmt, b.con.Id)

		q, err := b.queueDeclare(qName)
		if err != nil {
			return nil, err
		}

		if err := b.queueBind(q, q.Name, exchangeNames[i]); err != nil {
			return nil, err
		}

		qs = append(qs, q)
	}

	return qs, nil
}

func (b *Broker) initOutput() ([][]amqp.Queue, error) {
	exchangeName := b.con.OutputExchangeName
	qNames := b.con.OutputQueueNames
	qCopies := b.con.OutputCopies

	if err := b.exchangeDeclare(exchangeName, "direct"); err != nil {
		return nil, err
	}

	qs := make([][]amqp.Queue, 0, len(qNames)*len(qCopies))
	for i := range qNames {
		copyQs := make([]amqp.Queue, 0, len(qNames[i]))
		nameFmt := qNames[i] + "-%d"
		for id := range qCopies[i] {
			qName := fmt.Sprintf(nameFmt, id)

			q, err := b.queueDeclare(qName)
			if err != nil {
				return nil, err
			}

			if err := b.queueBind(q, q.Name, exchangeName); err != nil {
				return nil, err
			}

			copyQs = append(copyQs, q)
		}

		qs = append(qs, copyQs)
	}

	return qs, nil
}

func (b *Broker) exchangeDeclare(name string, kind string) error {
	return b.ch.ExchangeDeclare(
		name,
		kind,
		true,  // durable
		false, // auto-deleted
		false, // internal
		false, // no-wait
		nil,   // arguments
	)
}

func (b *Broker) queueDeclare(name string) (amqp.Queue, error) {
	return b.ch.QueueDeclare(
		name,
		false, // durable
		false, // delete when unused
		false, // exclusive
		false, // no-wait
		nil,   // arguments
	)
}

func (b *Broker) queueBind(q amqp.Queue, key string, exchangeName string) error {
	return b.ch.QueueBind(
		q.Name,
		key,
		exchangeName,
		false, // no-wait
		nil,   // args
	)
}

func (b *Broker) Consume(q amqp.Queue, consumer string) (<-chan amqp.Delivery, error) {
	return b.ch.Consume(
		q.Name,
		consumer,
		false, // auto-ack
		false, // exclusive
		false, // no-local
		false, // no-wait
		nil,   // args
	)
}

func (b *Broker) Publish(key string, body []byte) error {
	return b.ch.Publish(
		b.con.OutputExchangeName,
		key,
		false, // mandatory
		false, // immediate
		amqp.Publishing{
			ContentType: "application/octet-stream",
			Body:        body,
		})
}

func (b *Broker) PublishWithHeaders(key string, body []byte, headers amqp.Table) error {
	return b.ch.Publish(
		b.con.OutputExchangeName,
		key,
		false, // mandatory
		false, // immediate
		amqp.Publishing{
			ContentType: "application/octet-stream",
			Body:        body,
			Headers:     headers,
		})
}
